protobuf/protoc-gen-cruxclient/cruxclient_generator.cc

```diff
@@ -0,0 +1,426 @@
+  const char *str = str_start;
+  for (str = strpbrk(str, remove); str != nullptr;
+       str = strpbrk(str + 1, remove)) {
+    (*s)[str - str_start] = replacewith;
+  }
+}
+
+// The following function was copied from
+// google/protobuf/src/google/protobuf/compiler/cpp/cpp_helpers.cc
+
+string StripProto(const string &filename) {
+  if (HasSuffixString(filename, ".protodevel")) {
+    return StripSuffixString(filename, ".protodevel");
+  } else {
+    return StripSuffixString(filename, ".proto");
+  }
+}
+
+// The following 3 function was copied from
+// grpc/src/compiler/cpp_generator.cc
+
+string FilenameIdentifier(const string &filename) {
+  string result;
+  for (unsigned i = 0; i < filename.size(); i++) {
+    char c = filename[i];
+    if (isalnum(c)) {
+      result.push_back(c);
+    } else {
+      static char hex[] = "0123456789abcdef";
+      result.push_back('_');
+      result.push_back(hex[(c >> 4) & 0xf]);
+      result.push_back(hex[c & 0xf]);
+    }
+  }
+  return result;
+}
+
+inline string StringReplace(string str, const string &from, const string &to,
+                            bool replace_all) {
+  size_t pos = 0;
+
+  do {
+    pos = str.find(from, pos);
+    if (pos == string::npos) {
+      break;
+    }
+    str.replace(pos, from.length(), to);
+    pos += to.length();
+  } while (replace_all);
+
+  return str;
+}
+
+inline string StringReplace(string str, const string &from, const string &to) {
+  return StringReplace(str, from, to, true);
+}
+
+// The following 4 function was copied from
+// grpc/src/compiler/generator_helpers.h
+inline std::vector<string> tokenize(const string &input,
+                                    const string &delimiters) {
+  std::vector<string> tokens;
+  size_t pos, last_pos = 0;
+
+  for (;;) {
+    bool done = false;
+    pos = input.find_first_of(delimiters, last_pos);
+    if (pos == string::npos) {
+      done = true;
+      pos = input.length();
+    }
+
+    tokens.push_back(input.substr(last_pos, pos - last_pos));
+    if (done) return tokens;
+
+    last_pos = pos + 1;
+  }
+}
+
+// The following 3 function was copied from
+// grpc/src/compiler/cpp_generator_helpers.h
+
+inline string DotsToColons(const string &name) {
+  return StringReplace(name, ".", "::");
+}
+
+inline string DotsToUnderscores(const string &name) {
+  return StringReplace(name, ".", "_");
+}
+
+inline string ClassName(const Descriptor *descriptor, bool qualified) {
+  // Find "outer", the descriptor of the top-level message in which
+  // "descriptor" is embedded.
+  const Descriptor *outer = descriptor;
+  while (outer->containing_type() != NULL) outer = outer->containing_type();
+
+  const string &outer_name = outer->full_name();
+  string inner_name = descriptor->full_name().substr(outer_name.size());
+
+  if (qualified) {
+    return DotsToColons(outer_name) + DotsToUnderscores(inner_name);
+  } else {
+    return outer->name() + DotsToUnderscores(inner_name);
+  }
+}
+
```

Comment:
Can you add a comment saying custom crux generation code starts here.

Reply:
Done